test/pytest/mockwebserver.py

```diff
@@ -0,0 +1,492 @@
+    return web.Response(text=request.query_string.replace(' ', '+'))
+
+
+@routes.get(r'/cockpit/channel/{csrf_token}')
+async def cockpit_channel(request: web.Request) -> web.StreamResponse:
+    try:
+        transport = MultiplexTransport.transports[request.match_info['csrf_token']]
+    except KeyError:
+        return web.Response(status=404)
+
+    # Decode the request
+    try:
+        options = json.loads(binascii.a2b_base64(request.query_string))
+    except (json.JSONDecodeError, binascii.Error) as exc:
+        return web.Response(status=400, reason=f'Invalid query string {exc!s}')
+
+    binary = get_enum(options, 'binary', ['raw'], None) == 'raw'
+    websocket = request.headers.get('Upgrade', '').lower() == 'websocket'
+
+    # Open the channel, requesting data send to our queue
+    queue = asyncio.Queue[JsonObject | bytes]()
+    channel = transport.register_origin(ExternalChannelOrigin(queue.put_nowait))
+    transport.control_received({**options, 'command': 'open', 'channel': channel, 'flow-control': True})
+
+    # The first thing the channel sends back will be 'ready' or 'close'
+    open_result = await queue.get()
+    assert isinstance(open_result, Mapping)
+    if get_str(open_result, 'command') != 'ready':
+        return web.json_response(open_result, status=400, reason='Failed to open channel')
+
+    # Start streaming the result.
+    if websocket:
+        response: web.StreamResponse = web.WebSocketResponse()
+        await response.prepare(request)
+
+    else:
+        # Send the 'external' field back as the HTTP headers...
+        headers = {k: typechecked(v, str) for k, v in get_dict(options, 'external', {}).items()}
+
+        if 'Content-Type' not in headers:
+            headers['Content-Type'] = 'application/octet-stream' if binary else 'text/plain'
+
+        # ...plus this, if we have it.
+        if size_hint := get_int(open_result, 'size-hint', None):
+            headers['Content-Length'] = f'{size_hint}'
+
+        response = web.StreamResponse(status=200, headers=headers)
+        await response.prepare(request)
+
+        # Now, handle the data we receive
+        while item := await queue.get():
+            match item:
+                case Mapping():
+                    match get_str(item, 'command'):
+                        case 'ping':
+                            transport.control_received({**item, 'command': 'pong'})
+                        case 'close' | 'done':
+                            break
+
+                case bytes():
+                    await response.write(item)
+
+    return response
+
+
+@routes.get(r'/cockpit/socket')
+async def cockpit_socket(request: web.Request) -> web.WebSocketResponse:
+    ws = CockpitWebSocket()
+    request.app[websockets].add(ws)
+    await ws.communicate(request)
+    return ws
+
+
+@routes.get('/')
+async def index(_request: web.Request) -> web.Response:
+    cases = Path('qunit').rglob('test-*.html')
```

Comment:
Hmm, am I silly? There is no `qunit` path? So I am a bit confused what this lists.

Reply:
Like it's mostly under `base1`?

Reply:
esbuild puts the built bundles of pkg/base1/test* into qunit/*/test*.